File: release/ci-compile.rb
# compile itch for production environemnts

require_relative 'common'

module Itch
  def Itch.ci_compile
    say "Compiling #{app_name}"

    show_versions %w(npm node)

    ✓ npm_dep 'grunt', 'grunt-cli'
    ✓ npm 'install'

    say "Compiling JavaScript"
    ENV['NODE_ENV'] = 'production'
    ✓ grunt 'babel sass copy'

    say "Preparing stage"
    stage_path = 'stage'
    FileUtils.rm_rf stage_path
    FileUtils.mkdir_p stage_path

    say "Copying compiled code+assets..."
    FileUtils.cp_r 'app', stage_path
    FileUtils.cp_r 'node_modules', stage_path

    say "Generating custom package.json + environment"
    pkg = JSON.parse(File.read('package.json'))
    %w(name productName desktopName).each do |field|
      pkg[field] = app_name
    end
    env = {
      "name" => "production",
      "channel" => channel_name
    }
    File.write("#{stage_path}/package.json", JSON.pretty_generate(pkg))

    envjs = %Q{
module.exports = #{JSON.pretty_generate(env)}
}
    File.write("#{stage_path}/app/env.js", envjs)

    ✓ sh "tar cfz stage.tar.gz stage"
  end
end

Itch.ci_compile

File: release/ci-compile.js
#!/usr/bin/env node

// compile itch for production environemnts

const $ = require("./common");
const humanize = require("humanize-plus");

async function main() {
  $.say(`Preparing to compile ${$.appName()} ${$.buildVersion()}`);

  await $.showVersions(["npm", "node"]);

  $(await $.npm("install"));

  $.say("Wiping prefix...");
  $(await $.sh("rm -rf prefix"));
  $(await $.sh("mkdir -p prefix"));

  $.say("Compiling sources...");
  $(await $.sh("npm run compile"));

  $.say("Copying dist to prefix...");
  $(await $.sh("cp -rf dist prefix/"));

  $.say("Copying static resources to prefix...");
  $(await $.sh("mkdir prefix/src"));
  $(await $.sh("cp -rf src/static prefix/src"));

  $.say("Generating custom package.json...");
  const pkg = JSON.parse(await $.readFile("package.json"));
  for (const field of ["name", "productName", "desktopName"]) {
    pkg[field] = $.appName();
  }
  delete pkg.scripts.postinstall;
  pkg.version = $.buildVersion();
  const pkgContents = JSON.stringify(pkg, null, 2);
  await $.writeFile(`prefix/package.json`, pkgContents);

  $.say("Compressing prefix...");
  $(await $.sh("tar cf prefix.tar prefix"));

  const stats = await $.lstat("prefix.tar");
  $.say(`prefix.tar is ${humanize.fileSize(stats.size)}`);
}

main();
